flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptivebatch/DefaultVertexParallelismAndInputInfosDeciderTest.java

```diff
@@ -0,0 +1,523 @@
+
+    @Test
+    void testHavePointwiseEdges() {
+        final DefaultVertexParallelismAndInputInfosDecider decider =
+                createVertexParallelismAndInputInfosDecider(1, 10, 60L);
+
+        AllToAllBlockingResultInfo resultInfo1 =
+                createAllToAllBlockingResultInfo(
+                        new long[] {10L, 15L, 13L, 12L, 1L, 10L, 8L, 20L, 12L, 17L});
+        PointwiseBlockingResultInfo resultInfo2 =
+                createPointwiseBlockingResultInfo(
+                        new long[] {8L, 12L, 21L, 9L, 13L}, new long[] {7L, 19L, 13L, 14L, 5L});
+        ParallelismAndInputInfos parallelismAndInputInfos =
+                decider.decideParallelismAndInputInfosForVertex(
+                        Arrays.asList(resultInfo1, resultInfo2), -1);
+
+        assertThat(parallelismAndInputInfos.getParallelism()).isEqualTo(4);
+        assertThat(parallelismAndInputInfos.getJobVertexInputInfos()).hasSize(2);
+
+        checkAllToAllJobVertexInputInfo(
+                parallelismAndInputInfos.getJobVertexInputInfos().get(resultInfo1.getResultId()),
+                Arrays.asList(
+                        new IndexRange(0, 1),
+                        new IndexRange(2, 4),
+                        new IndexRange(5, 6),
+                        new IndexRange(7, 9)));
+        checkPointwiseJobVertexInputInfo(
+                parallelismAndInputInfos.getJobVertexInputInfos().get(resultInfo2.getResultId()),
+                Arrays.asList(
+                        new IndexRange(0, 0),
+                        new IndexRange(0, 0),
+                        new IndexRange(1, 1),
+                        new IndexRange(1, 1)),
+                Arrays.asList(
+                        new IndexRange(0, 1),
+                        new IndexRange(2, 4),
+                        new IndexRange(0, 1),
+                        new IndexRange(2, 4)));
+    }
+
+    @Test
+    void testParallelismAlreadyDecided() {
+        final DefaultVertexParallelismAndInputInfosDecider decider =
+                createVertexParallelismAndInputInfosDecider();
+
+        AllToAllBlockingResultInfo allToAllBlockingResultInfo =
+                createAllToAllBlockingResultInfo(
+                        new long[] {10L, 15L, 13L, 12L, 1L, 10L, 8L, 20L, 12L, 17L});
+        ParallelismAndInputInfos parallelismAndInputInfos =
+                decider.decideParallelismAndInputInfosForVertex(
+                        Collections.singletonList(allToAllBlockingResultInfo), 3);
+
+        assertThat(parallelismAndInputInfos.getParallelism()).isEqualTo(3);
+        assertThat(parallelismAndInputInfos.getJobVertexInputInfos()).hasSize(1);
+
+        checkAllToAllJobVertexInputInfo(
+                Iterables.getOnlyElement(
+                        parallelismAndInputInfos.getJobVertexInputInfos().values()),
+                Arrays.asList(new IndexRange(0, 2), new IndexRange(3, 5), new IndexRange(6, 9)));
+    }
+
+    @Test
+    void testSourceJobVertex() {
```

Comment:
Can we have a case to verify that when a vertex consumes a result multiple times, the result is correct?